### PR TITLE
[ONNX] Introduce ONNX reference evaluator for verification

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89947
* #89946
* __->__ #89808
* #89807

Reference evaluator requires ONNX >= 1.13. Running in CI is blocked by unable to bump onnx submodule version, like in #83201. Local tests pass.

### CODE REVIEW

File: torch/onnx/verification.py [hunk 1]
```diff
@@ -166,16 +187,33 @@ def _ort_session(
 
 
 @_beartype.beartype
-def _compare_ort_pytorch_outputs(
-    ort_outs: Union[Sequence[_NumericType], Sequence],
+def _onnx_reference_evaluator_session(model: Union[str, io.BytesIO]):
```
Comment (justinchuby) [inline torch/onnx/verification.py:190]: Return type

Comment (justinchuby) [reply torch/onnx/verification.py:190]: nvm

File: torch/onnx/verification.py [hunk 2]
```diff
@@ -29,6 +30,17 @@
...
 _NumericType = Union[Number, torch.Tensor, np.ndarray]
+_ModelType = Union[torch.nn.Module, torch.jit.ScriptModule]
+_InputArgsType = Union[torch.Tensor, Tuple[Any, ...]]
+_InputKwargsType = Mapping[str, Any]
+
+
+class OnnxBackend(enum.Enum):
+    """Enum class for ONNX backend used for export verification."""
+
+    ONNX = "ONNXReferenceEvaluator"
```
Comment (justinchuby) [inline torch/onnx/verification.py:41]: REFERENCE? Just a thought

File: test/onnx/test_verification.py [hunk 3]
```diff
@@ -114,8 +155,14 @@ def test_compare_ort_pytorch_outputs_raise_without_acceptable_error_percentage(
...
         with self.assertRaises(AssertionError):
-            verification._compare_ort_pytorch_outputs(
+            verification._compare_onnx_pytorch_outputs(
                 ort_outs,
                 pytorch_outs,
                 options,
             )
+
+
+common_utils.instantiate_parametrized_tests(TestVerification)
```
Comment (justinchuby) [inline test/onnx/test_verification.py:165]: I would use this as a decorator.

**Review by justinchuby — APPROVED** [review]
LGTM🚀

**Comment by titaiwangms** [comment]
> Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
> 
> * [[ONNX] Add repro export from `GraphInfo` #89947](https://github.com/pytorch/pytorch/pull/89947)
> * [[ONNX] Verification tool to find mismatch in model export #89946](https://github.com/pytorch/pytorch/pull/89946)
> * **->** [[ONNX] Introduce ONNX reference evaluator for verification #89808](https://github.com/pytorch/pytorch/pull/89808)
> * [[ONNX] Use `VerificationOptions` to wrap option arguments #89807](https://github.com/pytorch/pytorch/pull/89807)
> * [Small fix for `torch._C.Graph` type hint #89821](https://github.com/pytorch/pytorch/pull/89821)
> 
> Reference evaluator requires ONNX >= 1.13. Running in CI is blocked by #83201. Local tests pass.

I am not sure if #83201 includes all reference evaluator code, as I see some of "reference implementation" PR after Aug 10. I guess we will have to update #83201 to formal ONNX 1.13 release?

**Comment by BowenBao** [comment]
> > Reference evaluator requires ONNX >= 1.13. Running in CI is blocked by #83201. Local tests pass.
> 
> I am not sure if #83201 includes all reference evaluator code, as I see some of "reference implementation" PR after Aug 10. I guess we will have to update it to formal ONNX 1.13 release?

Yes you are right. I should write it as blocked by unable to bump onnx submodule version ..

**Comment by titaiwangms** [comment]
> Yes you are right. I should write it as blocked by unable to bump onnx submodule version ..

I am updating the ONNX version on that PR.

**Pull request closed without merging**